### DISCUSSION TITLE
how to do simple macros with parameters?

### DISCUSSION THREAD

**Discussion by PlasmaHH** [fralau/mkdocs-macros-plugin#232] [Q&A]
Hi, is it possible to do simple macros with parameters? There are wikis out there that have various syntax like `{macro:arg1=foo}` or `{{macro arg1="foo"}}`  or very similar ones and then the definition body would be like `Hello $arg1`  or similar and then the result would be `Hello foo`.

Is anything like that easily possible with simple macro definitions in yaml files? I know I could write python code for that but it seems quite some overhead for a simple text replacement, also it is much less accessible for all the people at the company that do just want to write text and not program in python...

**Comment by fralau** [comment]
Sure! Their syntax is similar to Python.

Just [follow the example](https://mkdocs-macros-plugin.readthedocs.io/en/latest/#full-example), and enjoy!

**Reply by PlasmaHH** [reply]
Sorry but can you point me to the specific example? I was unable to locate anything similar... As I said, I know already that I could write python code... Maybe a more "realistic" example would be something like

definition of such a macro be similar to:

`edit: [Edit](http://example.com/edit/{file}`

end then call it maybe with a syntax similar to

`{{ edit(file="xxx.md") }}`

and then the resulting expansion be like

``[Edit](http://example.com/edit/xxx.md`

**Reply by PlasmaHH** [reply]
So I wrote me a "macro" to do that replacement in the following way, maybe something like that could be part of the official plugin, I think it would be quite useful for format string style macros:

```python
import functools

class _KwDict(dict):
    """ A dictionary that will on missing keys return a string with the keyname instead """

    def __missing__(self,key):
        """
        @brief When the key was not found in the normal dict we return a string with its name in the appropriate brackets instead
        @param key The key
        """
        return f"{{{key}}}"


def format( s, *args, **kwargs ):

    try:
        return s.format(*args,**kwargs)
    except (TypeError,KeyError):
        d=_KwDict(kwargs)
        for i,a in enumerate(args):
            d[str(i)] = a
        try:
            return s.format_map(d)
        except ValueError:
            for k,v in d.items():
                s = s.replace(f"{{{k}}}",str(v))
            return s

def define_env(env):
    obsolete = []

    for k,v in env.variables.items():
        if( not isinstance(v,str) ):
            continue
        b = v.find("{")
        e = v.find("}")
        if( b != -1 and b < e ):
            func = functools.partial( format, v )
            env.macro(func,k)
            obsolete.append(k)
    for o in obsolete:
        env.variables.pop(o)
```

**Reply by fralau** [reply]
In theory you could call them as:

```
Bargain: {{ format("For only {%s:.2f} dollars!",price) }}
```

But I am not sure why you would do that, and what the code under `define_env()` is meant for? 🤔 


If you want to format text, why don't you first use the standard [Jinja2 `format` filter?](https://tedboy.github.io/jinja2/templ14.html#format). 

It comes out of the box and it is very easy to use:

```
Bargain:  For only {{ '%s:.2f' | format(price) }} dollars!
```

I believe that before going further in writing macros, you should first exploit standard Jinja2, and see how far you can go?

The [official doc on Jinka2](https://jinja.palletsprojects.com/en/3.0.x/templates/) uses HTML style text as an example, but it also works for Markdown.

**Reply by PlasmaHH** [reply]
I don't want to format some objects with format strings, I want to have wiki style macros that expand to some extra information with the passed one embedded. For example, defining one macro as

```yaml
lmgtfy: [{0}](https://www.google.com/search?q={0})
```

and then later in markdown do

```md
{{ lmgtfy("mkdcos") }}
```

To render a link to a google search for the passed term. So in case the URI one day changes, or you want to add paramterer everywhere or whatever, all you have to do is to change the definition.

> But I am not sure why you would do that, and what the code under define_env() is meant for? 🤔

The code is a (good enough for me, probably too simple for others) way to go through the macro plugins variables and if they contain a { before a } somewhere replace it by a function in the macro dict.


I am all ears how to do it easier (from the writer of the .md file and potentially maintianer of the macro definition list) with jinja templates only

**Reply by fralau** [reply]
I got it!

That should do it:

```python
def define_env(env):
    
    @env.macro
    def lgmtfy(search_item):
        return f"https://www.google.com/search?q={search_item}"
```

Then you could use, somewhere in your markdown:

```markdown
Here is the URL: {{ lmgtfy("mkdocs") }}
```

Then usually, you would have to use the hyperlink syntax so that it becomes clickable, like this:

```markdown
[Here is the URL]({{ lmgtfy("mkdocs") }})
```

**Reply by PlasmaHH** [reply]
My point here is to make the macro definition easy, alongside all the other macros in a single file. I don't want the people that write the documentation who have no idea of python to fiddle with the python code. 

It is also much more convenient for me to have all macro definitions in a single yaml file, as ~90% of them will not need to have that expansion be done. Also I plan one of the macro definition files to be autogenerated, it would be really convenient to use that instead having to auto generated python code.

**Reply by fralau** [reply]
MkDocs-Macros should give you the tools to do all that (puglets, YAML file), if you are willing to put in the homework.

**Reply by PlasmaHH** [reply]
Well I guess my python code from above is that homework then as I could not find any other way to make it possible to define parametrized macro in yaml
